### PR TITLE
Added clock_gettime, clock_getres, clock_settime, clock_getcpuclockid

### LINKED ISSUE #1275: Add support for clock_gettime/clock_getcpuclockid
I'm interested in adding those, if there are no concerns with that.

### PR DESCRIPTION
**Pull request by xonatius**
Picked up #1100 and added `clock_getcpuclockid` call as well. Credits to @kevinwern for the initial version.

https://www.man7.org/linux/man-pages/man2/clock_gettime.2.html
https://www.man7.org/linux/man-pages/man3/clock_getcpuclockid.3.html

Closes #1275

### CODE REVIEW

**Comment by xonatius** [comment]
@asomers @kamalmarhubi @posborne

Friendly ping for code review, thanks!

File: src/time.rs
Comment (asomers) [inline src/time.rs:26]: This function is supported by FreeBSD, too.

File: src/time.rs
Comment (asomers) [inline src/time.rs:35]: Since this method is basically the same as `std::time::Instant::now`, it would be good if the API could be similar.  Do you agree?  The `time` method could be renamed `now`, and we could add convertability between `TimeSpec` and `Instant`.

File: src/time.rs
Comment (xonatius) [reply src/time.rs:26]: Indeed, although it's not in rust-lang/libc bindings for FreeBSD. I'll try to add it and update the PR.

File: src/time.rs
Comment (xonatius) [reply src/time.rs:35]: Agreed. Wasn't sure if `now()` makes sense for pid-based clocks, but I guess it sort of does as we get the time on the pid-based clock at the current moment. I'll update the PR.

File: src/time.rs
Comment (xonatius) [reply src/time.rs:26]: [Added to more targets](https://github.com/rust-lang/libc/pull/1867) in libc waiting for a [version bump](https://github.com/rust-lang/libc/pull/1879) to update this PR.

File: src/time.rs
Comment (xonatius) [reply src/time.rs:35]: Looking more, it seems like `std::time::Instant` is pretty limited in its usage and seems to be designed to just get the current monotonically increasing time, rather than all purpose type to represent an arbitrary point in time (i.e. the only way to construct `std::time::Instant` is to call `Instant::now()`).

The other problem is that I don't think it makes sense to convert the time process has been running on CPU to a particular point in time since unix epoch. Looking on C `timespec` [documentation](https://en.cppreference.com/w/c/chrono/timespec) I think `TimeSpec` is closer to `std::time::Duration` rather than `std::time::Instant` (or even `std::time::SystemTime`), as it represents the interval rather than particular point in time.

If that makes sense, I can add convertability between `TimeSpec` and `Duration`.

Comment (asomers) [reply src/time.rs:35]: Hm, good point.  I agree.  Duration is more appropriate.

**Review by asomers — CHANGES_REQUESTED** [review]
Could you please fix the test failures?

**Comment by xonatius** [comment]
Looks like with [a recent change](https://github.com/rust-lang/libc/pull/1870) to libc to change some functions from unsafe to safe on linux platform, the usage of those function is now broken across platforms. I'll look into fixing those, probably the best way is to make functions safe on other platforms as well...

**Comment by asomers** [comment]
PR #1294 will fix the unused unsafe warnings.  However, I like your change better, because it's more finely tuned.  Would you consider opening a separate PR just for those?

**Comment by xonatius** [comment]
Looks like that PR is updated to be more finely tuned, so I guess a separate PR is no longer needed. I'll merge and update my changes once that PR is merged.

On a side note, once https://github.com/rust-lang/libc/pull/1891 gets into new release the `unsafe` blocks and `allow(unused_unsafe)` won't be needed anymore.

**Comment by xonatius** [comment]
Clicked the wrong button :/

**Comment by vi** [comment]
> On a side note, once rust-lang/libc#1891 gets into new release the unsafe blocks and allow(unused_unsafe) won't be needed anymore.

`allow(unused_unsafe)` with unsafe blocks can still be beneficial to extend the range of supported `libc` versions.

**Comment by xonatius** [comment]
> `allow(unused_unsafe)` with unsafe blocks can still be beneficial to extend the range of supported `libc` versions.

Sure, but that probably extends only by 1 version, as this change already bumps up the `libc` version to the latest one and can't be built without it.

**Comment by asomers** [comment]
You'll have to rebase; don't merge master into your branch.

**Comment by xonatius** [comment]
Done, let me know if you prefer me to squash changes into a single commit.

File: src/time.rs
Comment (asomers) [inline src/time.rs:19]: Should this use `libc_enum!`?

File: src/time.rs
Comment (asomers) [inline src/time.rs:206]: Better to link to the portable man page. https://pubs.opengroup.org/onlinepubs/7908799/xsh/clock_settime.html

File: src/time.rs
Comment (asomers) [inline src/time.rs:216]: Better to link to the portable man page. https://pubs.opengroup.org/onlinepubs/7908799/xsh/clock_settime.html

File: src/time.rs
Comment (asomers) [inline src/time.rs:226]: Better to link to the portable man page. https://pubs.opengroup.org/onlinepubs/7908799/xsh/clock_settime.html

File: src/time.rs
Comment (asomers) [inline src/time.rs:241]: Better to link to the portable man page: https://pubs.opengroup.org/onlinepubs/009695399/functions/clock_getcpuclockid.html

**Review by asomers — CHANGES_REQUESTED** [review]

File: src/time.rs
Comment (xonatius) [reply src/time.rs:19]: The enum won't work for `clockid_t` returned by `clock_getcpuclockid`, that's why I had to go with struct.

File: src/time.rs
Comment (xonatius) [reply src/time.rs:206]: Thanks, done!

File: src/time.rs
Comment (asomers) [reply src/time.rs:19]: You're worried that the OS may add a new clockid that Nix doesn't know about?  Ok, that's a fair reason.

**Comment by asomers** [comment]
LGTM!  Just squash the commits now.

**Comment by xonatius** [comment]
Done!

**Review by asomers — APPROVED** [review]
bors r+

**Pull request merged**
